htlcswitch/link.go

```diff
@@ -2106,6 +2326,28 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 			"ChannelPoint(%v): received error from peer: %v",
 			l.channel.ChannelPoint(), msg.Error(),
 		)
+
+	case *lnwire.Shutdown:
+		// We've received a Shutdown message from the remote peer.
+		// We'll set the shutdownReceived bool and cancel back any new
+		// HTLC's received after this point instead of forwarding them.
+		if l.shutdownReceived {
+			// Fail the link if a duplicate shutdown is received.
+			// This is also checked in peer.Brontide, but done here
+			// to properly stop the link.
+			l.fail(
+				LinkFailureError{code: ErrRemoteError},
+				"ChannelPoint(%v): received dupe shutdown "+
+					"from peer", l.channel.ChannelPoint(),
+			)
+			return
+		}
+
+		// Otherwise, set the shutdownReceived bool.
+		l.shutdownReceivedMtx.Lock()
+		l.shutdownReceived = true
```

Comment:
Don't we need to stash the actual message somewhere (since it contains the shutdown addr) to feed back to the main peer?